### PR TITLE
implementing TEP-0150 - `displayName` in `childReferences`

### PR DESCRIPTION
**Pull request by pritidesai**
# Changes

Implementing proposal TEP-0150.

Adding a functionality where the existing displayName can be used to configure unique names for each instance using the params from each combination such that the matrix instances in the Tekton Dashboard are rendered to distinguishable names.

Also, repurposing an existing name field as part of the matrix.include[].name. If specified, a fully resolved name field is available in the childReferences.

Fixes https://github.com/tektoncd/pipeline/issues/7082

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] [pre-commit](https://github.com/tektoncd/pipeline/blob/main/DEVELOPMENT.md#install-tools) Passed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
A fully resolved displayName is now available in childReferences along with the pipelineTaskName. This is mainly beneficial to parameterize and easily distinguish matrix instances of the task. 
```

### CODE REVIEW

File: docs/matrix.md [hunk 1]
```diff
@@ -155,6 +155,84 @@ Combinations generated
...
+
+Matrix creates multiple `taskRuns` with the same `pipelineTask`. Each `taskRun` has its unique combination `params` based
+on the `matrix` specifications. These `params` can now be surfaced and used to configure unique name of each `matrix`
+instance such that it is easier to distinguish all the instances based on their inputs.
+
+```yaml
+pipelineSpec:
+  tasks:
+    - name: platforms-and-browsers
+      displayName: "Platforms and Browsers: $(params.platform) and $(params.browser)"
```
Comment (AlanGreene) [inline docs/matrix.md:168]: Should this be `matrix.params.platform` and `matrix.params.browser`?

It looks like this wasn't included in the final version of the proposal that was merged, but I wonder if that was an oversight as it looked like it was agreed in the review feedback.

i.e. that there would be matrix-specific variables for the substitution to differentiate from other params, e.g.: https://github.com/tektoncd/community/pull/1122#discussion_r1452386634

This would be important to allow adding support for pipelineTask param substitution in future.

Maybe I missed some other discussion that changed this plan?

Comment (pritidesai) [reply docs/matrix.md:168]: thanks @AlanGreene for following up!

We had a discussion in the API WG. A keyword `matrix` in `params.platform` is not needed as the `displayName` is now part of `childReferences` irrespective of whether a task is a `matrix` or not.

With this approach, the pipelineTask param substitution in displayName is also done in this PR.

Do you think this could cause any issues while the clients are trying to rely on this change?

Comment (pritidesai) [reply docs/matrix.md:168]: > Is it possible to have a matrix param with the same name as a pipelineTask / implicitly propagated param? 

Hey @AlanGreene, matrix params has to be named same as the task params and implicitly mapped to the task params. In the example of platform and browsers, both those parameters are mapped to those params of the referenced task.

It is possible that the pipeline author can include matrix params which does not exist in the referenced task and hence will be missing from the `taskRun`. But that can create other issues before getting to the `displayName`. If not then the displayName cannot be resolved as they will not be present in the `taskRun`.

Comment (pritidesai) [reply docs/matrix.md:168]: I think I understand now what you are referring to 😥 

It is not allowed, we have a validation in place to make sure the same param either exist in the matrix or as pipelineTask param.

```
    tasks:
      - name: platforms-and-browsers
        displayName: "Platform: $(params.platform) with Browser: $(params.browser)"
        params:
          - name: platform
            value: linux1
        matrix:
          params:
            - name: platform
              value:
                - linux
```

results in:

```
 k create -f examples/v1/pipelineruns/beta/pipelinerun-with-matrix.yaml 
task.tekton.dev/platform-browsers created
Error from server (BadRequest): error when creating "examples/v1/pipelineruns/beta/pipelinerun-with-matrix.yaml": admission webhook "validation.webhook.pipeline.tekton.dev" denied the request: validation failed: expected exactly one, got both: spec.pipelineSpec.tasks[0].matrix[platform], spec.pipelineSpec.tasks[0].params[platform]
```

Comment (AlanGreene) [reply docs/matrix.md:168]: OK that's good to know, however there may still be an issue with propagated params, see for example:

```yaml
apiVersion: tekton.dev/v1
kind: PipelineRun
metadata:
  generateName: pr-echo-
spec:
  params:
    - name: HELLO
      value: "Hello World!"
  pipelineSpec:
    tasks:
      - name: echo-hello
        displayName: Message is $(params.HELLO)
        matrix:
          params:
            - name: HELLO
              value:
                - hi
                - hey
        taskSpec:
          steps:
            - name: echo
              image: ubuntu
              script: |
                #!/usr/bin/env bash
                echo "$(params.HELLO)"
```

This is accepted and executed by the controller today, however it does not produce the expected result. The matrix params are correctly applied to the TaskRun params, however the params in the step definition have already been replaced by the pipeline params.

This seems like a bug, or at least an undocumented limitation, with the current matrix implementation.

How would this scenario be handled with the new displayName processing?

Comment (afrittoli) [reply docs/matrix.md:168]: Conflicts between task/matrix parameters and pipeline parameters are indeed a possibility. 
If anything, I believe we should give precedence to local (matrix) parameters in case of naming conflict, as it is more likely that the author would like to use those in the `displayName` rather than a fixed pipeline parameter.

That would be an acceptable compromise for me, which would allow us to avoid introducing the `matrix` prefix

Comment (AlanGreene) [reply docs/matrix.md:168]: Matrix params taking precedence makes sense 👍 Of course it should be consistent between the displayName and actual usage in the steps.

Do we need a separate issue for the bug illustrated by the example above where param substitution in the steps is NOT using the matrix params?

Comment (AlanGreene) [reply docs/matrix.md:168]: Posting here to record offline discussion
- option 1 sounds good to me
- and we'll file a bug to track the bad substitution for taskSpec

Comment (pritidesai) [reply docs/matrix.md:168]: Hey Alan, I have created an issue here - https://github.com/tektoncd/pipeline/issues/7704

**Comment by pritidesai** [comment]
> The display name here (and in the next ones) seems to be related to how a specific matrix combination was extended through the matrix.include section. Some are only extended with common-package so they all that name. the the common-package include did not exists, would they have an empty displayName?

Thanks @afrittoli! That is correct. The `displayName` in `childReferences` will have the `matrix.include[].name` only for the combinations which were added as part of that `include` otherwise the `displayName` is not initialized. For the names from multiple include sections, we are concatenating them with a `space`. Any specific comment/feedback on that?

**Comment by afrittoli** [comment]
> > The display name here (and in the next ones) seems to be related to how a specific matrix combination was extended through the matrix.include section. Some are only extended with common-package so they all that name. the the common-package include did not exists, would they have an empty displayName?
> 
> Thanks @afrittoli! That is correct. The `displayName` in `childReferences` will have the `matrix.include[].name` only for the combinations which were added as part of that `include` otherwise the `displayName` is not initialized. For the names from multiple include sections, we are concatenating them with a `space`. Any specific comment/feedback on that?

Thanks @pritidesai. I think concatenation makes sense. As I mentioned, I find the current matrix API a bit hard to parse, but that's not something for this to address. Given the current matrix API, the proposed display name makes sense. Perhaps we should have one case in the test list where display name remains empty to check that case too?

**Comment by pritidesai** [comment]
> Thanks @pritidesai. I think concatenation makes sense. As I mentioned, I find the current matrix API a bit hard to parse, but that's not something for this to address. Given the current matrix API, the proposed display name makes sense. Perhaps we should have one case in the test list where display name remains empty to check that case too?

We have a variety of tests written under [pipelinerunstate_test.go](https://github.com/tektoncd/pipeline/pull/7683/files#diff-22f0dfe65e555d6833c559a7a03613fa9cfec74ddf0962109a8b7ed810ccc850R3705) which include the case where the displayName remains empty when include params does not make it as part of the combinations:

https://github.com/tektoncd/pipeline/pull/7683/files#diff-22f0dfe65e555d6833c559a7a03613fa9cfec74ddf0962109a8b7ed810ccc850R4014-R4021

Let me know if we want to add more tests. Thanks!

**Review by AverageMarcus — COMMENTED** [review]
/lgtm

**Review by AlanGreene — COMMENTED** [review]
> Pipeline WG: @AlanGreene @AverageMarcus PTAL again since we are hoping to get this merged this month.

The latest updates to the docs and examples look good based on our discussions and previous feedback, and as far as I can tell the test updates make sense too.

However, I'm not familiar enough with the codebase to determine whether the implementation has any potential impact on other features / functionality, in particular the changes in `ApplyReplacements`. A Pipelines maintainer or contributor would need to review and comment on that.

**Comment by pritidesai** [comment]
@tektoncd/core-maintainers please help review this PR. Thanks in advance!

**Comment by pritidesai** [comment]
This is a very simple change in adding human readable name in the childReferences. As an end user, we are waiting for this feature. Will appreciate if we can get the community to review on this PR. Thanks in advance!

File: pkg/apis/pipeline/v1beta1/pipelinerun_types.go [hunk 2]
```diff
@@ -432,6 +432,9 @@ type ChildStatusReference struct {
 	runtime.TypeMeta `json:",inline"`
 	// Name is the name of the TaskRun or Run this is referencing.
 	Name string `json:"name,omitempty"`
+	// DisplayName is a user-facing name of the pipelineTask that may be
+	// used to populate a UI.
+	DisplayName string `json:"displayName,omitempty"`
```
Comment (chitrangpatel) [inline pkg/apis/pipeline/v1beta1/pipelinerun_types.go:437]: Does there need to be a conversion between v1beta1 and v1?

Comment (chitrangpatel) [reply pkg/apis/pipeline/v1beta1/pipelinerun_types.go:437]: Otherwise lgtm. Happy to approve as soon as this is resolved.

Comment (pritidesai) [reply pkg/apis/pipeline/v1beta1/pipelinerun_types.go:437]: thank you @chitrangpatel for the review, appreciate it! 🙏 

I am not sure about it. I have added the conversion [here](https://github.com/tektoncd/pipeline/compare/39daaf77ba155e10a994bb300855f0536205ef02..a9f4b619c0b57ea28c931c531bc872f65ad5893e), does it address the conversion? 🤔

File: docs/matrix.md (hunk 1 above)
Comment (afrittoli) [reply docs/matrix.md:168]: Marking this conversation as resolved - thanks @pritidesai for issue #7704

File: pkg/reconciler/pipelinerun/resources/pipelinerunstate_test.go [hunk 3]
```diff
@@ -3642,6 +3702,418 @@ func TestPipelineRunState_GetChildReferences(t *testing.T) {
...
+				Name:             "matrixed-task-run-1",
+				DisplayName:      "task running foo and random",
+				PipelineTaskName: "matrixed-task",
+			}, {
+				TypeMeta: runtime.TypeMeta{
+					APIVersion: "tekton.dev/v1",
+					Kind:       "TaskRun",
+				},
+				Name:             "matrixed-task-run-2",
+				PipelineTaskName: "matrixed-task",
```
Comment (afrittoli) [inline pkg/reconciler/pipelinerun/resources/pipelinerunstate_test.go:4108]: +1

File: pkg/reconciler/pipelinerun/resources/pipelinerunstate.go [hunk 4]
```diff
@@ -240,8 +242,53 @@ func (facts *PipelineRunFacts) GetChildReferences() []v1.ChildStatusReference {
 	return childRefs
 }
 
+func (t *ResolvedPipelineTask) getDisplayName(customRun *v1beta1.CustomRun, taskRun *v1.TaskRun, c v1.ChildStatusReference) v1.ChildStatusReference {
```
Comment (afrittoli) [inline pkg/reconciler/pipelinerun/resources/pipelinerunstate.go:245]: NIT: I find the name/structure of this function slightly confusing - it does not return the display name, instead it sets it on a `ChildStatusReference` that must be passed as input. 

Since the `ChildStatusReference` is not used to build the displayName, I thought it would be better to have a `getDisplayName` function that returns a string - any particular reason for having it this way?
